entity-framework/core/miscellaneous/nativeaot-and-precompiled-queries.md

```diff
@@ -0,0 +1,110 @@
+
+C# interceptors are currently an experimental feature, and require a special opt-in in your project file:
+
+```xml
+<PropertyGroup>
+  <InterceptorsNamespaces>$(InterceptorsPreviewNamespaces);Microsoft.EntityFrameworkCore.GeneratedInterceptors</InterceptorsNamespaces>
+</PropertyGroup>
+```
+
+At this point, you're ready to precompile your LINQ queries, and generate the [compiled model](xref:core/performance/advanced-performance-topics#compiled-models) that they depend on. Make sure that you have at least version 9.0 of the EF tools (`dotnet tool list -g`), and then execute the following:
+
+```console
+dotnet ef dbcontext optimize --precompile-queries --nativeaot
+```
+
+You're now ready to publish your EF NativeAOT application:
+
+```console
+dotnet publish -r linux-arm64 -c Release
+```
+
+This shows publishing a NativeAOT publishing for Linux running on ARM64; [consult this catalog](/dotnet/core/rid-catalog) to find your runtime identifier.
+
+Due to the way C# interceptors work, any change in the application source invalidates them and requires repeating the above process. As a result, interceptor generation and actual publishing aren't expected to happen in the inner loop, as the developer is working on code; instead, both `dotnet ef dbcontext optimize` and `dotnet publish` can be executed in a publishing/deployment workflow, in a CI/CD system.
+
+> [!NOTE]
+> Publishing currently reports a number of trimming and NativeAOT warnings, meaning that your application isn't fully guaranteed to run properly. This is expected given the current experimental state of NativeAOT support; the final, non-experimental feature will report no warnings.
+
+## Limitations
+
+### Dynamic queries are not supported
+
+Query precompilation performs static analysis of your source code, identifying EF LINQ queries and generating C# interceptors for them. LINQ allows expressing highly dynamic queries, where LINQ operators are composed based on arbitrary conditions; such queries unfortunately cannot be statically analyzed, and are currently unsupported. Consider the following example:
+
+```c#
+IAsyncEnumerable<Blog> GetBlogs(BlogContext context, bool applyFilter)
+{
+    IQueryable<Blog> query = context.Blogs.OrderBy(b => b.Id);
+
+    if (applyFilter)
+    {
+        query = query.Where(b => b.Name != "foo");
+    }
+
+    return query.AsAsyncEnumerable();
+}
+```
+
+The above query is split across several statements, and dynamically composes the `Where` operator based on an external parameter; such queries cannot be precompiled. However, it is sometimes possible to rewrite such dynamic queries as multiple non-dynamic queries:
```

Comment:
Such queries will (very certainly) never be recompiled, since that would require analyzing arbitrary code patterns statically - that's just not possible in the general case... We can still support such dynamic queries by compiling them at runtime - like today - but that's not precompilation...

Reply:
1. Never say never.
2. We might in the future recognize _some_ patterns.
3. For the text it is really not important. Feel free to ignore this.

Reply:
OK :)

In the general case, I definitely feel confident to say "never" here - since IQueryables can be passed around between functions (and even assemblies), and arbitrary .NET code (with loops what what-no) can be used to compose them, that just ain't happening. I'm not sure I see us investing effort in recognizing specific patterns, but who knows... In any case, as you say this is just a bit of docs, we can always change it later.